test/test_ops.py

```diff
@@ -712,14 +729,19 @@ def _create_tensors_with_iou(self, N, iou_thresh):
 
     @pytest.mark.parametrize("iou", (0.2, 0.5, 0.8))
     @pytest.mark.parametrize("seed", range(10))
-    def test_nms_ref(self, iou, seed):
+    def test_nms_ref(self, iou, seed, dtype=torch.float):
         torch.random.manual_seed(seed)
         err_msg = "NMS incompatible between CPU and reference implementation for IoU={}"
         boxes, scores = self._create_tensors_with_iou(1000, iou)
         keep_ref = self._reference_nms(boxes, scores, iou)
         keep = ops.nms(boxes, scores, iou)
         torch.testing.assert_close(keep, keep_ref, msg=err_msg.format(iou))
 
+        if dtype == torch.bfloat16:
+            keep_ref_float = ops.nms(boxes.to(dtype).float(), scores.to(dtype).float(), iou)
+            keep_dtype = ops.nms(boxes.to(dtype), scores.to(dtype), iou)
+            torch.testing.assert_close(keep_ref_float, keep_dtype)
```

Comment:
Thanks for the update and sorry for the noise - shouldn't we instead assert that we get equivalent results when autocast is on and when it's off? e.g. something like

```py
    @pytest.mark.parametrize("iou", (0.2, 0.5, 0.8))
    @pytest.mark.parametrize("seed", range(10))
    def test_autocast_cpu(self, iou, seed):
        torch.random.manual_seed(seed)
        boxes, scores = self._create_tensors_with_iou(1000, iou)
        keep = ops.nms(boxes, scores, iou)
        with torch.cuda.amp.autocast(dtype=torch.bfloat16):
            keep_autocast = ops.nms(boxes, scores, iou)
        torch.testing.assert_close(keep, keep_autocast)
```

The test for roi_align calls `test_forward()` which has a slightly different logic where instead we check our implem against a reference, both with autocast. Unfortunately that same test seems to fail for nms, so maybe we can just use the snippet above. WDYT?

Reply:
Thanks for the suggestion. Here are two concerns:
1. Since nms is not in the white list of autocast, simply running the above UT won't be able to expose the issue that this PR is trying to fix, since the input to nms will still be fp32 in the above snippet. Instead, we need to provide an input in BF16 to nms under autocast, so that without this PR, it will throw `RuntimeError: "nms_kernel" not implemented for 'BFloat16'`, while with this PR, autocast will convert the input back to FP32 so that it could run successfully.
2. A modified version following item 1 will be like this:
  ```python
      @pytest.mark.parametrize("iou", (0.2, 0.5, 0.8))
      @pytest.mark.parametrize("seed", range(10))
      def test_autocast_cpu(self, iou, seed):
          torch.random.manual_seed(seed)
          boxes, scores = self._create_tensors_with_iou(1000, iou)
          keep = ops.nms(boxes, scores, iou)
          with torch.cpu.amp.autocast(dtype=torch.bfloat16):
              keep_autocast = ops.nms(boxes.to(torch.bfloat16), scores.to(torch.bfloat16), iou)
          torch.testing.assert_close(keep, keep_autocast)
  ```
  The issue with this version is that, since `boxes.to(torch.bfloat16)` and `scores.to(torch.bfloat16)` has converted data from high precision to low precision, when directly comparing the result `keep_autocast` and `keep`, we'll meet:
  `AssertionError: The values for attribute 'shape' do not match: torch.Size([434]) != torch.Size([436]).`. 
  I used `boxes.to(dtype).float()` in the current implementation [here](https://github.com/chunyuan-w/vision/blob/a792605b5e47852dbb9870773fd9d1edf3442500/test/test_ops.py#L741) to simulate the process of converting from high precision to low precision and convert it back to float to be the reference result to compare with the autocast result.